tests/unit/test-class-wc-payments-payment-request-button-handler.php

```diff
@@ -549,4 +549,27 @@ public function test_get_product_data_returns_the_same_as_build_display_items_wi
 			'Failed asserting total amount are the same for get_product_data and build_display_items'
 		);
 	}
+
+	public function test_filter_cart_needs_shipping_address_returns_false() {
+		$mock_pr = $this->getMockBuilder( WC_Payments_Payment_Request_Button_Handler::class )
+			->setConstructorArgs( [ $this->mock_wcpay_account, $this->mock_wcpay_gateway, $this->express_checkout_helper ] )
+			->setMethods( [ 'has_subscription_product', 'get_shipping_method_count' ] )
+			->getMock();
+
+		$mock_pr->method( 'has_subscription_product' )->willReturn( true );
+		$mock_pr->method( 'get_shipping_method_count' )->willReturn( 0 );
+
+		$this->assertFalse( $mock_pr->filter_cart_needs_shipping_address( true ) );
+	}
+
+	public function test_filter_cart_needs_shipping_address_returns_true() {
+		$mock_pr = $this->getMockBuilder( WC_Payments_Payment_Request_Button_Handler::class )
```

Comment:
While we figured out how to get rid of `get_shipping_method_count` mock, we still are mocking `has_subscription_product` which forces us to mock the unit under test. This is not ideal but has less negative impact compared to creating a wrapper class around the shipping method count (because the mock is only in the test), but still it'd be great to avoid this. I haven't tried to control `WC_Subscriptions_Cart::cart_contains_subscription` because shipping count took me some time, but I'd guess something under `tests/unit/subscriptions` can potentially be helpful. I'm leaving this improvement as a nice-to-have (not must).

Reply:
>because shipping count took me some time

This is exactly why I didn't spend time trying to control the cart subscription. Turns out its quite easy with `WC_Subscriptions_Cart::set_cart_contains_subscription( true )`.